geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/executor/sortedset/ZStoreExecutor.java

```diff
@@ -0,0 +1,133 @@
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package org.apache.geode.redis.internal.executor.sortedset;
+
+import static org.apache.geode.redis.internal.RedisConstants.ERROR_KEY_REQUIRED;
+import static org.apache.geode.redis.internal.RedisConstants.ERROR_SYNTAX;
+import static org.apache.geode.redis.internal.RedisConstants.ERROR_WEIGHT_NOT_A_FLOAT;
+import static org.apache.geode.redis.internal.RedisConstants.ERROR_WRONG_SLOT;
+import static org.apache.geode.redis.internal.netty.Coder.narrowLongToInt;
+import static org.apache.geode.redis.internal.netty.Coder.toUpperCaseBytes;
+import static org.apache.geode.redis.internal.netty.StringBytesGlossary.bAGGREGATE;
+import static org.apache.geode.redis.internal.netty.StringBytesGlossary.bWEIGHTS;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.ListIterator;
+
+import org.apache.geode.redis.internal.data.RedisKey;
+import org.apache.geode.redis.internal.executor.AbstractExecutor;
+import org.apache.geode.redis.internal.executor.RedisResponse;
+import org.apache.geode.redis.internal.netty.Coder;
+import org.apache.geode.redis.internal.netty.Command;
+import org.apache.geode.redis.internal.netty.ExecutionHandlerContext;
+
+public abstract class ZStoreExecutor extends AbstractExecutor {
+
+  @Override
+  public RedisResponse executeCommand(Command command, ExecutionHandlerContext context) {
+    List<byte[]> commandElements = command.getProcessedCommand();
+    ListIterator<byte[]> argIterator = commandElements.listIterator();
+    RedisResponse syntaxErrorResponse = RedisResponse.error(ERROR_SYNTAX);
+
+    // Skip command and destination key
+    argIterator.next();
+    argIterator.next();
+
+    // get the number of keys
+    int numKeys;
+    try {
+      numKeys = narrowLongToInt(Coder.bytesToLong(argIterator.next()));
+      if (numKeys > commandElements.size() - 2) {
+        return syntaxErrorResponse;
+      } else if (numKeys <= 0) {
+        return RedisResponse.error(ERROR_KEY_REQUIRED);
+      }
+    } catch (NumberFormatException ex) {
+      return syntaxErrorResponse;
+    }
+
+    List<ZKeyWeight> keyWeights = new ArrayList<>(numKeys);
+    ZAggregator aggregator = ZAggregator.SUM;
+    byte[] argument;
+
+    // get all the keys
+    for (int i = 0; i < numKeys; i++) {
+      if (!argIterator.hasNext()) {
+        return syntaxErrorResponse;
+      }
+      argument = argIterator.next();
+      if (Arrays.equals(toUpperCaseBytes(argument), bWEIGHTS)
+          || Arrays.equals(toUpperCaseBytes(argument), bAGGREGATE)) {
+        return syntaxErrorResponse;
+      }
```

Comment:
This behaviour is incorrect. "WEIGHTS" and "AGGREGATE" are both valid Redis key names. The below test fails with our implementation, but passes on native Redis:
```
@Test
  public void shouldNotReturnError_givenKeysNamedWeightsOrAggregate() {
    String weights = "WEIGHTS";
    jedis.zadd(weights, 1, "member");
    String aggregate = "AGGREGATE";
    jedis.zadd(aggregate, 1, "member");
    jedis.zinterstore(weights, weights, weights);
    jedis.zinterstore(aggregate, aggregate, aggregate);
  }
```

Reply:
good point, thank you.